persistence/secrets.go

```diff
@@ -0,0 +1,281 @@
+			}
+		}
+	}
+
+	return outputSecrets
+}
+
+// Saves the given secret to the agent db
+func SaveSecret(db *bolt.DB, secretName string, msDefId string, msDefVers string, secretToSave PersistedServiceSecret) error {
+	secretToSaveAll := PersistedServiceSecrets{MsDefId: msDefId, MsDefOrg: secretToSave.SvcOrgid, MsDefUrl: secretToSave.SvcUrl, MsDefVers: msDefVers, SecretsMap: map[string]PersistedServiceSecret{}}
+	if allSecsForServiceInDB, err := FindAllSecretsForMS(db, msDefId); err != nil {
+		return fmt.Errorf("Failed to get all secrets for microservice %v. Error was: %v", msDefId, err)
+	} else if allSecsForServiceInDB != nil {
+		secretToSaveAll = *allSecsForServiceInDB
+	}
+
+	if secretToSave.TimeCreated == 0 {
+		secretToSave.TimeCreated = uint64(time.Now().Unix())
+	}
+
+	secretToSave.TimeLastUpdated = uint64(time.Now().Unix())
+
+	if mergedSec, ok := secretToSaveAll.SecretsMap[secretName]; ok {
+		mergedSec.AgreementIds = append(mergedSec.AgreementIds, secretToSave.AgreementIds...)
+		secretToSaveAll.SecretsMap[secretName] = mergedSec
+	} else {
+		secretToSaveAll.SecretsMap[secretName] = secretToSave
+	}
+
+	return SaveAllServiceSecrets(db, msDefId, secretToSaveAll)
+}
+
+func SaveAllServiceSecrets(db *bolt.DB, msDefId string, secretToSaveAll PersistedServiceSecrets) error {
+	writeErr := db.Update(func(tx *bolt.Tx) error {
+		bucket, err := tx.CreateBucketIfNotExists([]byte(SECRETS))
+		if err != nil {
+			return err
+		}
+
+		if serial, err := json.Marshal(secretToSaveAll); err != nil {
+			return fmt.Errorf("Failed to serialize secrets: Error: %v", err)
+		} else {
+			return bucket.Put([]byte(msDefId), serial)
+		}
+	})
+
+	return writeErr
+}
+
+// Gets the secret from the database, no error returned if none is found in the db
+func FindAllSecretsForMS(db *bolt.DB, msDefId string) (*PersistedServiceSecrets, error) {
+	var psecretRec *PersistedServiceSecrets
+	readErr := db.View(func(tx *bolt.Tx) error {
+
+		if b := tx.Bucket([]byte(SECRETS)); b != nil {
+			s := b.Get([]byte(msDefId))
+			if s != nil {
+				secretRec := PersistedServiceSecrets{}
+				if err := json.Unmarshal(s, &secretRec); err != nil {
+					glog.Errorf("Unable to deserialize service secret db record: %v. Error: %v", msDefId, err)
+					return err
+				} else {
+					psecretRec = &secretRec
+				}
+			}
+		}
+		return nil
+	})
+
+	if readErr != nil {
+		return nil, readErr
+	} else {
+		return psecretRec, nil
+	}
+}
+
+// Find a particular secret, if a version range is provided it must match the exact range on the secret, if a specific version is given return the first matching secret range
+func FindSingleSecretForService(db *bolt.DB, secName string, msDefId string) (*PersistedServiceSecret, error) {
+	allSec, err := FindAllSecretsForMS(db, msDefId)
+	if err != nil {
+		return nil, err
+	}
+
+	if allSec != nil {
+		retSec := allSec.SecretsMap[msDefId]
```

Comment:
Should this be `retSec := allSec.SecretsMap[secName]`?

Reply:
Yes, thanks